R/bounds.r

```diff
@@ -1,3 +1,14 @@
+# If needed, converts input from classes that lack division and multiplication
+# to classes that can be scaled.
+# @param x vector of values to convert, if needed
+adapt_input_class <- function(x) {
```

Comment:
`scales::rescale` mainly depends on elementary arithmetic operations (addition, subtraction, multiplication and division). Coercing to numeric is a good way to define multiplication and division operations for `POSIXct` objects, but it is not the way to go for all R objects. For instance `integer64` numbers, from the `bit64` package, define their own arithmetic operators to prevent loss of precision.

If the coercion to numeric is done, there will be a warning when using some integer64 objects:

```
x <- bit64::as.integer64(c(2^50, 2^54+1, 2^60+2))
from <- range(x, na.rm = TRUE)
# I need to compute the `from` range here due to an unrelated bit64 bug triggered
# when range(..., finite = TRUE) is used. (this bug has already been
# reported to the bit64 maintainer)

# No warning given now:
scales::rescale(x = x, from = from)
# [1] 0.00000000 0.01466276 1.00000000

# If we force coercion to numeric there is a warning:
scales::rescale(x = as.numeric(x), from = as.numeric(from))
# Warning messages:
# 1: In as.double.integer64(from) :
#   integer precision lost while converting to double
# 2: In as.double.integer64(x) :
#   integer precision lost while converting to double
# [1] 0.00000000 0.01466276 1.00000000
```

I wanted to provide an example were the `as.numeric` approach is wrong, but `scales::rescale` calls `scales::zero_range` that has a default tolerance too large for `integer64` data. I could work towards letting the user give a custom tolerance value, but for the sake of the explanation you can set `tol` to zero manually in the `zero_range` definition in `r/bounds.r` and rebuild, in case you want to reproduce this example:

```
x <- bit64::as.integer64(2^60) + 1:3 
# x has three consecutive numbers, we expect the result to be c(0, 0.5, 1)
x_range <- range(x, na.rm = TRUE)
scales::rescale(x = x, from = x_range) # Correct result
# [1] 0.0 0.5 1.0
scales::rescale(x = as.numeric(x), from = as.numeric(x_range)) # loss of precision
# [1] 0.5 0.5 0.5
```

Even if we have tolerance issues in this case, I prefer to work towards making `scales` work with other objects, and therefore not coercing everything to `numeric`.

But you have the last word, though :smiley: